### PR TITLE
Cache decompositions in low rank integrator

### PR DESCRIPTION
**Pull request by p-pietro**
This PR caches the QR decomposition or pseudoinverse in `MESolveLowRankIntegrator`, so that these expensive operations are performed only once per evaluation of the field. This leads to a significant performance improvement when many jump operators are present.

To accomplish this while keeping AD compatibility, the dependency on `lineax` has been dropped, while reimplementing the relevant rules directly.

### CODE REVIEW

File: dynamiqs/integrators/core/low_rank_integrator.py [hunk 1]
```diff
@@ -22,20 +23,74 @@
...
+    primals: tuple[Any, ...], tangents: tuple[Any, ...]
+) -> tuple[Array, Array]:
+    A, cache, B = primals
+    dA, _dcache, dB = tangents
+    x = _qr_solve_cached(A, cache, B)
+    # Pseudoinverse JVP formula (independent columns):
+    # dX = A^+ (dB - dA X + (A^+)^H (dA^H (B - A X))).
+    res = B - A @ x
+    dvec = dB - dA @ x + _qr_pinv_h_apply(cache, dA.conj().T @ res)
+    dx = _qr_pinv_apply(cache, dvec)
```
Comment (Nicolas-Lepage) [inline dynamiqs/integrators/core/low_rank_integrator.py]: this has the same name as `import diffrax as dx`, even if this causes no issue in this file we should rename it

File: dynamiqs/integrators/core/low_rank_integrator.py [hunk 2]
```diff
@@ -22,20 +23,74 @@
...
+@jax.custom_jvp
+def _cholesky_solve_cached(A: Array, A_pinv: Array, B: Array) -> Array:
+    del A
     return A_pinv @ B
 
 
+@_cholesky_solve_cached.defjvp
+def _cholesky_solve_cached_jvp(
+    primals: tuple[Any, ...], tangents: tuple[Any, ...]
+) -> tuple[Array, Array]:
```
Comment (Nicolas-Lepage) [inline dynamiqs/integrators/core/low_rank_integrator.py:99]: i think the typing can be tightened here with something like 

```
def _cholesky_solve_cached_jvp(
    primals: tuple[Array, Array, Array],
    tangents: tuple[Array, Array, Array],
) -> tuple[Array, Array]: ...
```

File: dynamiqs/integrators/core/low_rank_integrator.py [hunk 3]
```diff
@@ -22,20 +23,74 @@
...
+@jax.custom_jvp
+def _qr_solve_cached(A: Array, cache: tuple[Array, Array], B: Array) -> Array:
+    del A
+    return _qr_pinv_apply(cache, B)
+
+
+@_qr_solve_cached.defjvp
+def _qr_solve_cached_jvp(
+    primals: tuple[Any, ...], tangents: tuple[Any, ...]
+) -> tuple[Array, Array]:
```
Comment (Nicolas-Lepage) [inline dynamiqs/integrators/core/low_rank_integrator.py:74]: I think the typing can be tightened here: 

```
def _qr_solve_cached_jvp(
    primals: tuple[Array, tuple[Array, Array], Array],
    tangents: tuple[Array, tuple[Array, Array], Array],
) -> tuple[Array, Array]: ...
```

**Review by Nicolas-Lepage — CHANGES_REQUESTED** [review]
Hi @p-pietro! Thanks for the contribution, and sorry for taking a bit of time to review :)

The method looks good, I have a few comments before merging.

Style (non-scientific):
- `dx` is used as a local variable in both JVP rules, shadowing `import diffrax as dx`. Harmless here, but please rename it.
- Could you rename some of the shorthand notations , e.g. `res`, `dvec`, `pinv_h`, `Lm`... 
-`Any` can be replaced with stricter typing. 

tests : 
The existing low-rank tests do cover gradient accuracy against analytic solutions, but only for the default linear solver. 
- Can you arametrize `test_correctness` and `test_gradient` over `LinearSolver.QR` and `LinearSolver.CHOLESKY`

AD :
We're adding higher-order differentiation support to the library and will want to extend it to the low-rank integrator eventually. The cached rules are correct at first order but not at second. could you add a TODO comment to later fix compatibility ?

**Comment by p-pietro** [comment]
Hello @Nicolas-Lepage, thanks for the comments! I should have now addressed all of them.

**Review by Nicolas-Lepage — APPROVED** [review]
Thanks @p-pietro for the changes! Everything looks good to me.

@gautierronan, would you like to take a look as well before we merge?

**Review by gautierronan — APPROVED** [review]
LGTM! Thanks @p-pietro
